File: lib/python/opentoken/src/main/opentoken/io/csv/token_csv_writer.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import csv
import logging
import os
from typing import Dict

from opentoken.io.token_writer import TokenWriter
from opentoken.processor.token_constants import TokenConstants


logger = logging.getLogger(__name__)


class TokenCSVWriter(TokenWriter):
    """Writes decrypted tokens to a CSV file.

    Parity notes (Java `TokenCSVWriter`):
    - Java uses a simple BufferedWriter and manual comma separation.
    - We use `csv.DictWriter` for safer quoting; order and header names match.
    - Added an explicit flush after each write to approximate Java's immediate write semantics.
    Output columns: RuleId, Token, RecordId
    """

    def __init__(self, file_path: str):
        """
        Initialize the class with the output file in CSV format.

        Args:
            file_path: The output file path.

        Raises:
            IOError: If an I/O error occurs.
        """
        self.file_path = file_path

        # Create directory if it doesn't exist
        os.makedirs(os.path.dirname(file_path) if os.path.dirname(file_path) else '.', exist_ok=True)

        self.file_handle = open(file_path, 'w', newline='', encoding='utf-8')
        self.csv_writer = csv.DictWriter(
            self.file_handle,
            fieldnames=[TokenConstants.RULE_ID, TokenConstants.TOKEN, TokenConstants.RECORD_ID],
            lineterminator='\n'
        )
        self.csv_writer.writeheader()

    def write_token(self, data: Dict[str, str]) -> None:
        """Write a token row to the CSV file.

        Args:
            data: Dict containing RuleId, Token, RecordId (missing keys default to empty string).
        """
        try:
            self.csv_writer.writerow({
                TokenConstants.RULE_ID: data.get(TokenConstants.RULE_ID, ''),
                TokenConstants.TOKEN: data.get(TokenConstants.TOKEN, ''),
                TokenConstants.RECORD_ID: data.get(TokenConstants.RECORD_ID, '')
            })
            # Flush to keep behavior closer to Java's line-by-line writes
            self.file_handle.flush()
        except IOError as e:
            logger.error(f"Error in writing to CSV file: {e}")
            raise

    def close(self):
        """Close the file handle."""
        if self.file_handle:
            self.file_handle.close()

    def __enter__(self):
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        self.close()

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/io/csv/TokenCSVWriter.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.csv;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.io.TokenWriter;
import com.truveta.opentoken.processor.TokenConstants;

/**
 * Writes decrypted tokens to a CSV file.
 * Output columns: RuleId, Token, RecordId
 */
public class TokenCSVWriter implements TokenWriter {
    private static final Logger logger = LoggerFactory.getLogger(TokenCSVWriter.class);

    private final BufferedWriter writer;

    /**
     * Initialize the class with the output file in CSV format.
     * 
     * @param filePath The output file path.
     * @throws IOException If an I/O error occurs.
     */
    public TokenCSVWriter(String filePath) throws IOException {
        // Create directory if it doesn't exist
        Path path = Paths.get(filePath);
        Path parentDir = path.getParent();
        if (parentDir != null) {
            Files.createDirectories(parentDir);
        }

        this.writer = new BufferedWriter(new FileWriter(filePath));

        // Write header with LF line ending to match Python output
        writer.write(TokenConstants.RULE_ID + "," + TokenConstants.TOKEN + "," + TokenConstants.RECORD_ID);
        writer.write('\n');
    }

    /**
     * Write a token row to the CSV file.
     * 
     * @param data A map containing RuleId, Token, and RecordId.
     * @throws IOException If an I/O error occurs.
     */
    @Override
    public void writeToken(Map<String, String> data) throws IOException {
        String ruleId = data.getOrDefault(TokenConstants.RULE_ID, "");
        String token = data.getOrDefault(TokenConstants.TOKEN, "");
        String recordId = data.getOrDefault(TokenConstants.RECORD_ID, "");

        writer.write(String.format("%s,%s,%s", ruleId, token, recordId));
        writer.write('\n'); // Use LF line ending to match Python output
    }

    @Override
    public void close() throws IOException {
        if (writer != null) {
            writer.close();
        }
    }
}
